networkx/linalg/algebraicconnectivity.py

```diff
@@ -244,7 +244,7 @@ def P(x):
             W -= (W.T * X * X.T).T
             project(W)
             # Compute the diagonal of P * L * P as a Jacobi preconditioner.
-            D = L.diagonal()
+            D = L.diagonal().astype(float)
             D += 2. * (asarray(X) * asarray(W)).sum(axis=1)
```

Comment:
@dschult I believe the `D` in line 174 is not relevant in this case. The `D` in `lambda x: D * x` will use whatever value exists in the outside scope at the time of calling (due to late binding). So since `solver` is called in a `while` loop that includes a changing value of `D`, the lambda function will use a different `D` each time. @ysitu's suggestion would indeed make it clearer and we can probably make that change when we make `L` a float dtype matrix.